gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/DagProcessingEngineTest.java

```diff
@@ -0,0 +1,161 @@
+    config = configBuilder.build();
+
+    // Constructing TopologySpecMap.
+    Map<URI, TopologySpec> topologySpecMap = new HashMap<>();
+    String specExecInstance = "mySpecExecutor";
+    TopologySpec topologySpec = DagTestUtils.buildNaiveTopologySpec(specExecInstance);
+    URI specExecURI = new URI(specExecInstance);
+    topologySpecMap.put(specExecURI, topologySpec);
+    MostlyMySqlDagManagementStateStore dagManagementStateStore = new MostlyMySqlDagManagementStateStore(config, null);
+    dagManagementStateStore.setTopologySpecMap(topologySpecMap);
+    this.dagManagementTaskStream =
+        new DagManagementTaskStreamImpl(config, Optional.empty(), dagManagementStateStore);
+    this.dagManagementTaskStream.setActive(true);
+    this.dagProcFactory = new DagProcFactory();
+    this.dagProcEngineThread = new DagProcessingEngine.DagProcEngineThread(
+        this.dagManagementTaskStream, this.dagProcFactory, dagManagementStateStore);
+    this.dagTaskStream = spy(new MockedDagTaskStream());
+    this.dagProcessingEngine =
+        new DagProcessingEngine(config, dagTaskStream, this.dagProcFactory, dagManagementStateStore);
+  }
+
+  static class MockedDagTaskStream implements DagTaskStream {
+    public static final int MAX_NUM_OF_TASKS = 10;
+    volatile int i=0;
+
+    @Override
+    public boolean hasNext() {
+      return true;
+    }
+
+    @Override
+    public synchronized DagTask next() {
+      i++;
+      if (i <= MAX_NUM_OF_TASKS) {
+        return new MockedDagTask(new DagActionStore.DagAction("fg-" + i, "fn-" + i, "1234" + i, DagActionStore.FlowActionType.LAUNCH), null);
+      } else {
+        throw new RuntimeException("Max num of tasks reached");
+      }
+    }
+  }
+
+  static class MockedDagTask extends DagTask<MockedDagProc> {
+
+    public MockedDagTask(DagActionStore.DagAction dagAction, MultiActiveLeaseArbiter.LeaseObtainedStatus leaseObtainedStatus) {
+      super(dagAction, leaseObtainedStatus);
+    }
+
+    @Override
+    public MockedDagProc host(DagTaskVisitor<MockedDagProc> visitor) {
+      return new MockedDagProc();
+    }
+
+    @Override
+    public boolean conclude() throws IOException {
+      return false;
+    }
+  }
+
+  static class MockedDagProc extends DagProc<Void, Void> {
+
+    @Override
+    protected Void initialize(DagManagementStateStore dagManagementStateStore) {
+      return null;
+    }
+
+    @Override
+    protected Void act(DagManagementStateStore dagManagementStateStore, Void state) {
+      return null;
+    }
+
+    @Override
+    protected void sendNotification(Void result, EventSubmitter eventSubmitter) {
+    }
+
+    @Override
+    protected void commit(DagManagementStateStore dagManagementStateStore, Void result) {
+    }
+  }
+
+  // This tests verifies that
+  @Test
+  public void dagProcessingTest() throws InterruptedException, TimeoutException {
+    int expectedNumOfInvocations = MockedDagTaskStream.MAX_NUM_OF_TASKS + ServiceConfigKeys.DEFAULT_NUM_DAG_PROC_THREADS;
```

Comment:
is it the sum because there are numTasks returned and then each thread additionally blocks (infinitely) to wait for more?  if so, that's unlikely to be immediately clear to a maintainer so deserves a comment

Reply:
actually don't we expect 2x that number of invocations, given that we have both hasNext() and next()?